icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+        numBreaks++;
+    }
+
+    featureList->removeAllElements();
+    delete featureList;
+}
+
+void CjMLBreakEngine::initElementList(UnicodeString& inString, UVector* elementList, UErrorCode& status) const {
+    UChar32 w1, w2, w3, w4, w5, w6;
+    UnicodeString b1, b2, b3, b4, b5, b6;
+    w1 = INVALID;
+    w2 = INVALID;
+    w3 = inString.char32At(0);
+    w4 = inString.char32At(1);
+    w5 = 2 < inString.length() ? inString.char32At(2) : INVALID;
+    w6 = 3 < inString.length() ? inString.char32At(3) : INVALID;
+
+    b1 = UnicodeString(INVALID);
+    b2 = UnicodeString(INVALID);
+    b3 = getUnicodeBlockIndex(w3, status);
+    b4 = getUnicodeBlockIndex(w4, status);
+    b5 = getUnicodeBlockIndex(w5, status);
+    b6 = getUnicodeBlockIndex(w6, status);
+
+    elementList->addElement(new Element(w1, b1), status);
+    elementList->addElement(new Element(w2, b2), status);
+    elementList->addElement(new Element(w3, b3), status);
+    elementList->addElement(new Element(w4, b4), status);
+    elementList->addElement(new Element(w5, b5), status);
+    elementList->addElement(new Element(w6, b6), status);
+}
+
+/**
+ * Convert an integer, positive or negative, to a character string radix 10.
+ */
+static char* itoa(int32_t i, char* buf) {
+    char* result = buf;
+    // Handle negative
+    if (i < 0) {
+        *buf++ = '-';
+        i = -i;
+    }
+    // Output digits in reverse order
+    char* p = buf;
+    do {
+        *p++ = (char)('0' + (i % 10));
+        i /= 10;
+    } while (i);
+    *p-- = 0;
+    // Reverse the string
+    while (buf < p) {
+        char c = *buf;
+        *buf++ = *p;
+        *p-- = c;
+    }
+
+    return result;
+}
+
+UnicodeString CjMLBreakEngine::getUnicodeBlockIndex(UChar32 ch, UErrorCode& status) const{
+    U_ASSERT(fUnicodeBlock->size() > 0);
+    if (U_FAILURE(status)) {
+        return UnicodeString();
+    }
+    if (ch == INVALID) {
+        return UnicodeString(INVALID);
+    }
+    int32_t start = 0;
+    int32_t end = fUnicodeBlock->size();
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        if (ch < fUnicodeBlock->elementAti(mid)) {
+            end = mid;
+        }
+        else {
+            start = mid + 1;
+        }
+    }
+    //Add '0' to fill up three digits
+    char buff[32];
+    itoa(start, buff);
+    UnicodeString result(buff);
+    if (result.length() < 3) {
+        int32_t preZeroCount = 3 - result.length();
+        for (int i = 0; i < preZeroCount; i++) {
+            result.insert(0, '0');
+        }
+    }
+    return result;
+}
+
+void CjMLBreakEngine::loadMLparams(UErrorCode& error) {
+    negativeSum = 0;
+    const char *tags[] = { "modelKeys", "modelValues", "unicode_blocks" };
+    ResourceBundle ja(U_ICUDATA_BRKITR, "ja", error);
+    UnicodeString key;
+    int32_t size = 0;
+    int32_t index = 0;
+    const int32_t *value;
+
+    if (U_SUCCESS(error)) {
+        ResourceBundle keyBundle = ja.get(tags[0], error);
```

Comment:
Why do we put "modelKeys", "modelValues", "unicode_blocks" into a const char *tags[] then ref to them as tags[0], tags[1] and tags[2] in the code? I think this introduce another level of readability issue. also, could you write a block of comment before CjMLBreakEngine::loadMLparams to show the layout of the resources?

Reply:
Done.